### PR TITLE
[RISCV] fix RISCVPushPopOptimizer pass

### PR DESCRIPTION
**Pull request by daniilavdeev**
RISCVPushPopOptimizer pass didn't suggest that CFI instructions can be inserted between cm.pop and ret and couldn't apply optimization in these cases. The patch fix it and allows the pass to remove CFI instructions and combine cm.pop and ret into the one instruction: cm.popret or cm.popretz.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVPushPopOptimizer.cpp [hunk 1]
```diff
@@ -45,10 +45,28 @@ char RISCVPushPopOpt::ID = 0;
...
+static IterT nextNoDebugNoCFIInst(const IterT It, const IterT End) {
+  return std::find_if_not(std::next(It), End, [](const auto &Inst) {
+    return Inst.isDebugInstr() || Inst.isCFIInstruction() ||
+           Inst.isPseudoProbe();
+  });
+}
+
+static void eraseCFIInst(const MachineBasicBlock::iterator Begin,
+                         const MachineBasicBlock::iterator End) {
+  std::vector<std::reference_wrapper<llvm::MachineInstr>> CFIInstrs;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVPushPopOptimizer.cpp:58]: Why do we need to copy to a vector?

Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVPushPopOptimizer.cpp]: addressed

**Review by michaelmaitland — COMMENTED** [review]
test case?

**Comment by dlav-sc** [comment]
> test case?

This patch supports https://github.com/llvm/llvm-project/pull/110234, it doesn't really matter alone. I'm going to split https://github.com/llvm/llvm-project/pull/110234 into several PRs and after that this one will be a part of stacked PRs I guess.

**Comment by dlav-sc** [comment]
Actual version: https://github.com/llvm/llvm-project/pull/110813

**Comment by michaelmaitland** [comment]
> Actual version: #110813

Is this patch a duplicate? If so, we should close one of them.

**Comment by dlav-sc** [comment]
> we should close one of them.

yep

**Pull request closed without merging**